In [6]:
from sklearn.metrics.pairwise import linear_kernel
import re

In [32]:
import os
files = os.listdir("../corpus")
corpus = []

for file in files:
    dat = ""
    with open("../corpus/" + file, "r") as f:
        data = f.read()
        dat = re.sub("(<\S+>)", "" ,data)
        dat = re.sub("(^https?:\/\/\S+)", "", dat)
        dat = dat.replace(">", "")
        corpus.append(dat)
    with open("../preprocessed_corpus/" + file, "w") as f:
        f.write(dat)
    

In [38]:
len(corpus)

603

In [60]:
corpus[0]

'\n\n                    Trang chủ  \nCHUYÊN MỤC BỆNH HỌC  \nTiêu hóa - Gan mật  \nViêm trực tràng: Nguyên nhân, triệu chứng và cách phòng ngừa\nViêm trực tràng: Nguyên nhân, triệu chứng và cách phòng ngừa\nViêm trực tràng là bệnh lý có thể gặp phải ở nhiều lứa tuổi, kể cả ở trẻ em. Trong trường hợp cấp tính, người bệnh có thể hồi phục trong vòng vài tuần. Các vấn đề mạn tính có thể cần thời gian lâu hơn, bao gồm cả đều trị y tế và thay đổi lối sống. \nMục lục\nViêm trực tràng là gì?\nGiai đoạn viêm trực tràng\n1. Viêm trực tràng cấp tính\n2. Viêm trực tràng mạn tính\n Các triệu chứng viêm trực tràng là gì?\nNguyên nhân nào gây ra viêm trực tràng\nChẩn đoán viêm trực tràng\nBiến chứng thường gặp của bệnh viêm trực tràng\nĐiều trị bệnh viêm trực tràng\n1. Điều trị viêm trực tràng do nhiễm trùng\n2. Điều trị viêm trực tràng do xạ trị\n3. Điều trị viêm trực tràng do bệnh viêm ruột\n4. Phẫu thuật điều trị\n Phương pháp phòng ngừa viêm trực tràng như thế nào?\nViêm trực tràng là gì? Trực tr

Tạo mô hình TF-IDF 

In [39]:
from sklearn.feature_extraction.text import TfidfVectorizer
tf = TfidfVectorizer(analyzer='word', ngram_range=(1,3), min_df = 0, stop_words = 'english')

In [40]:
tfidf_matrix =  tf.fit_transform(corpus)
feature_names = tf.get_feature_names() 

/usr/local/lib/python3.9/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


In [44]:
feature_names[1:5]

['00 8211', '00 8211 00', '00 diop', '00 diop cận']

In [41]:
len(feature_names)

874363

Ma trận tf-idf có kích thước 603 x 874363 - mỗi một hàng biểu diễn một bệnh, mỗi một cột biểu diễn một cụm từ:

Lấy ra tập các từ và TF-IDF của nó 

In [42]:
dense = tfidf_matrix.todense()
episode = dense[0].tolist()[0]
phrase_scores = [pair for pair in zip(range(0, len(episode)), episode) if pair[1] > 0]

In [45]:
len(phrase_scores)

4082

In [43]:
sorted(phrase_scores, key=lambda t: t[1] * -1)[:5]

[(750296, 0.46654903646831747),
 (750297, 0.46654903646831747),
 (702980, 0.338611635324247),
 (675777, 0.3241058247536746),
 (702848, 0.18547401659770082)]

In [66]:
question = "Hương đang mang thai và lo lắng mình có thể gặp phải rau tiền đạo. Hương có thể kiểm tra phát hiện bệnh này từ tuần thứ mấy của thai kỳ?,A. Tuần 10,B.Tuần 20,C. Tuần 30,D. Tuần 40,,"
question

'Hương đang mang thai và lo lắng mình có thể gặp phải rau tiền đạo. Hương có thể kiểm tra phát hiện bệnh này từ tuần thứ mấy của thai kỳ?,A. Tuần 10,B.Tuần 20,C. Tuần 30,D. Tuần 40,,'

In [52]:
question_vec = tf.transform([question]).todense()

In [58]:
linear_kernel(question_vec, dense).argmax()

/usr/local/lib/python3.9/dist-packages/sklearn/utils/validation.py:593: FutureWarning: np.matrix usage is deprecated in 1.0 and will raise a TypeError in 1.2. Please convert to a numpy array with np.asarray. For more information see: https://numpy.org/doc/stable/reference/generated/numpy.matrix.html
  warnings.warn(
/usr/local/lib/python3.9/dist-packages/sklearn/utils/validation.py:593: FutureWarning: np.matrix usage is deprecated in 1.0 and will raise a TypeError in 1.2. Please convert to a numpy array with np.asarray. For more information see: https://numpy.org/doc/stable/reference/generated/numpy.matrix.html
  warnings.warn(


471

In [59]:
corpus[471]

'\n\n                    Trang chủ  \nCHUYÊN MỤC BỆNH HỌC  \nSản - Phụ khoa  \nRau tiền đạo: Nguyên nhân, dấu hiệu và cách phòng ngừa\nRau tiền đạo: Nguyên nhân, dấu hiệu và cách phòng ngừa\nRau tiền đạo là một trong những biến chứng thai kỳ nguy hiểm đe dọa tính mạng mẹ bầu và thai nhi với tỷ lệ mắc phải 1/200 trường hợp phụ nữ mang thai. Vậy dấu hiệu rau tiền đạo là gì để mẹ bầu sớm nhận biết và điều trị hiệu quả?  \nMục lục\nRau tiền đạo là gì?\nCác dạng rau tiền đạo thường gặp\nNguyên nhân gây ra rau tiền đạo\nDấu hiệu rau tiền đạo thường gặp\nRau tiền đạo có nguy hiểm không?\nĐối với thai phụ\nĐối với thai nhi\n Bao nhiêu tuần thì biết được rau tiền đạo?\nPhương pháp chẩn đoán rau tiền đạo\nĐiều trị rau tiền đạo như thế nào?\nĐối với rau tiền đạo khi chưa chuyển dạ\nĐối với rau tiền đạo khi chuyển dạ\nĐối với rau tiền đạo biến chứng rau cài răng lược\n Phòng ngừa rau tiền đạo\nRau tiền đạo là gì? Rau tiền đạo là hiện tượng rau thai nằm ở vị trí thấp nhất của tử cung, làm che mất m

BM25

In [61]:
class BM25:

    def __init__(self, k1=1.5, b=0.75):
        self.b = b
        self.k1 = k1

    def fit(self, corpus):
        """
        Fit the various statistics that are required to calculate BM25 ranking
        score using the corpus given.

        Parameters
        ----------
        corpus : list[list[str]]
            Each element in the list represents a document, and each document
            is a list of the terms.

        Returns
        -------
        self
        """
        tf = []
        df = {}
        idf = {}
        doc_len = []
        corpus_size = 0
        for document in corpus:
            corpus_size += 1
            doc_len.append(len(document))

            # compute tf (term frequency) per document
            frequencies = {}
            for term in document:
                term_count = frequencies.get(term, 0) + 1
                frequencies[term] = term_count

            tf.append(frequencies)

            # compute df (document frequency) per term
            for term, _ in frequencies.items():
                df_count = df.get(term, 0) + 1
                df[term] = df_count

        for term, freq in df.items():
            idf[term] = math.log(1 + (corpus_size - freq + 0.5) / (freq + 0.5))

        self.tf_ = tf
        self.df_ = df
        self.idf_ = idf
        self.doc_len_ = doc_len
        self.corpus_ = corpus
        self.corpus_size_ = corpus_size
        self.avg_doc_len_ = sum(doc_len) / corpus_size
        return self

    def search(self, query):
        scores = [self._score(query, index) for index in range(self.corpus_size_)]
        return scores

    def _score(self, query, index):
        score = 0.0

        doc_len = self.doc_len_[index]
        frequencies = self.tf_[index]
        for term in query:
            if term not in frequencies:
                continue

            freq = frequencies[term]
            numerator = self.idf_[term] * freq * (self.k1 + 1)
            denominator = freq + self.k1 * (1 - self.b + self.b * doc_len / self.avg_doc_len_)
            score += (numerator / denominator)

        return score

In [70]:
docs = corpus[471].split("\n")
texts = [
    [word for word in document.lower().split()]
    for document in docs
]

In [71]:
import math
bm25 = BM25()
bm25.fit(texts)

In [76]:
import numpy as np
limit = 3


query = question.split()

scores = bm25.search(query)
scores_index = np.argsort(scores)
scores_index = scores_index[::-1]

print(np.array([docs[i] for i in scores_index])[:limit])


['Dấu hiệu rau tiền đạo thường gặp Các triệu chứng của rau tiền đạo khác nhau tùy vào thể lâm sàng và mức độ nặng, nhẹ của bệnh. Nhìn chung, thai phụ có thể sớm nhận biết rau nếu gặp phải các triệu chứng này khi mang thai là: (3)  Bị xuất huyết âm đạo bất thường (máu có màu đỏ tươi, có thể lẫn cục máu) nhưng không hề gây cảm giác đau đớn, thường gặp ở 3 tháng cuối thai kỳ. Tình trạng xuất huyết âm đạo có thể từ mức độ nhẹ đến nặng, chảy máu ít hoặc nhiều, có thể tự hết mà không cần điều trị. Tuy nhiên, triệu chứng có thể tái phát với mức độ xuất huyết tăng dần. Một vài trường hợp thai phụ bị chảy máu đi kèm cơn co thắt.   '
 'Rau tiền đạo có nguy hiểm không? Rau tiền đạo nếu không được phát hiện sớm và can thiệp điều trị hiệu quả có thể gây biến chứng băng huyết sau sinh và trong thai kỳ, đe dọa tính mạng thai phụ lẫn thai nhi. Cụ thể là: '
 'Bao nhiêu tuần thì biết được rau tiền đạo? Bác sĩ Nguyễn Đức Hinh cho biết, hầu hết các trường hợp rau tiền đạo có thể được phát hiện sớm từ tuần

In [75]:
from sentence_transformers import SentenceTransformer

ModuleNotFoundError: No module named 'sentence_transformers'